app/src/main/java/com/lagradost/cloudstream3/utils/BiometricAuthenticator.kt

```diff
@@ -0,0 +1,177 @@
+            executor,
+            object : BiometricPrompt.AuthenticationCallback() {
+
+                override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
+                    super.onAuthenticationError(errorCode, errString)
+                    showToast("$errString", LENGTH_SHORT)
+                    Log.i(TAG, "$errorCode")
+                    activity.finish()
+                    failedAttempts++
+                    if (failedAttempts >= MAX_FAILED_ATTEMPTS) {
+                        failedAttempts = 0
+                        activity.finish()
+                    } else {
+                        failedAttempts = 0
+                        MainActivity().finish()
+                    }
+                }
+
+                override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
+                    super.onAuthenticationSucceeded(result)
+                    failedAttempts = 0
+                    authCallback?.onAuthenticationSuccess()
+                }
+
+                override fun onAuthenticationFailed() {
+                    super.onAuthenticationFailed()
+                    failedAttempts++
+                    if (failedAttempts >= MAX_FAILED_ATTEMPTS) {
+                        failedAttempts = 0
+                        activity.finish()
+                    }
+                }
+            })
+    }
+
+    private fun initBiometricPrompt(
+        activity: Activity,
+        title: Int,
+        setDeviceCred: Boolean,
+    ) {
+        val description = activity.getString(R.string.biometric_prompt_description)
+        if (setDeviceCred) {
+            // For API level > 30, Newer API setAllowedAuthenticators is used
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+
+                val authFlag = DEVICE_CREDENTIAL or BIOMETRIC_WEAK or BIOMETRIC_STRONG
+                promptInfo = BiometricPrompt.PromptInfo.Builder()
+                    .setTitle(activity.getString(title))
+                    .setDescription(description)
+                    .setAllowedAuthenticators(authFlag)
+                    .build()
+            } else {
+                @Suppress("DEPRECATION")
+                promptInfo = BiometricPrompt.PromptInfo.Builder()
+                    .setTitle(activity.getString(title))
+                    .setDescription(description)
+                    .setDeviceCredentialAllowed(true)
+                    .build()
+            }
+        } else {
+            // fallback for A12+ when both fingerprint & Face unlock is absent but PIN is set
+            @Suppress("DEPRECATION")
+            promptInfo = BiometricPrompt.PromptInfo.Builder()
+                .setTitle(activity.getString(title))
+                .setDescription(description)
+                .setDeviceCredentialAllowed(true)
+                .build()
+        }
+    }
+
+    private fun isBiometricHardWareAvailable(): Boolean {
+        var result = false
```

Comment:
I'd personally much prefer inlining this instead of using vars, but that is just a nitpick.

Reply:
current implementation responses robust truthy and falsy values(may look big but is good)